buckets/redis/bucket.go

```diff
@@ -56,32 +54,30 @@ func (a *abstractBucket) Take(ctx context.Context, requested int64, maxWaitTime
 		strconv.FormatInt(requested, 10), strconv.FormatInt(maxWaitTime.Nanoseconds(), 10),
 		maxIdleTimeMillis, a.maxDebtNanos}
 
-	var waitTime time.Duration
-	var err error
-
 	client := a.factory.Client().(*redis.Client)
 	res := a.takeFromRedis(ctx, client, args)
-	switch waitTimeNanos := res.Val().(type) {
+	if err := res.Err(); err != nil {
+		if isRedisClientClosedError(err) {
+			fmt.Println("reconnecting")
+			a.factory.handleConnectionFailure(client)
+		}
+		return 0, true, errors.Wrap(err, "failed to take token from redis bucket")
+	}
+
+	var waitTime time.Duration
+	switch val := res.Val().(type) {
 	case int64:
-		waitTime = time.Nanosecond * time.Duration(waitTimeNanos)
-		break
+		waitTime = time.Nanosecond * time.Duration(val)
 	default:
-		err = res.Err()
-		if unknownCloseError(err) {
-			logging.Printf("Unknown response '%v' of type %T. Full result %+v",
-				waitTimeNanos, waitTimeNanos, res)
-		}
-		// Handle connection failure
-		a.factory.handleConnectionFailure(client)
-		return 0, false
+		return 0, true, errors.Errorf("unknown response of type %[1]T: %[1]v", val)
```

Comment:
Do the `%`s need to be escaped?

Reply:
escaped?

Reply:
I'm not sure what you mean, the formatting directive works correctly: https://play.golang.org/p/EV-hmAUWPFt